### PR TITLE
Multi GPU with PEFT on LLM

### PR DESCRIPTION
**Pull request by avishniakov**
This PR brings the multi-GPU DDP showcase to the PEFT training.
It is using the function helper from the core and the other changes are mostly related to Accelerate specifically.

ZenML core companion PR: https://github.com/zenml-io/zenml/pull/2746 [  blocking merging of this one ]

### CODE REVIEW

**Comment by avishniakov** [comment]
@schustmi @htahir1 you are optional reviewers, just in case you have interest 🙂

**Review by htahir1 — COMMENTED** [review]
I havnt given this a fair shake but in general what is the better way of doing this vs using subprocess :-D Any ideas?

**Comment by htahir1** [comment]
@avishniakov Personally I feel they are quite unstable and unreliable... a better way would to use the internal library and do this in code right?

**Comment by avishniakov** [comment]
> @avishniakov Personally I feel they are quite unstable and unreliable... a better way would to use the internal library and do this in code right?

In theory, we can hack around the `accelerate.command.launch` module, but in this module they still will call `subprocess.Popen` for you, so you cannot get away from subprocessing anyway. I will explore how we can use the module directly.

File: llm-lora-finetuning/utils/cuda.py [hunk 1]
```diff
@@ -18,9 +18,13 @@
...
+from zenml.logger import get_logger
+
+logger = get_logger(__name__)
 
 
 def cleanup_memory() -> None:
     """Clean up GPU memory."""
+    logger.info("Cleaning up GPU memory...")
     while gc.collect():
         torch.cuda.empty_cache()
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply llm-lora-finetuning/utils/cuda.py]: Yeah, this was added. I will put something in README also.

**Comment by avishniakov** [comment]
I somewhat heavily reworked how the preparation of the functions was done in this project. This is tightly coupled with the changes on the ZenML side.

Looking forward to some conceptual feedback. There are definitely a few weak points:
- Cache invalidation mechanism due to the use of the external function is far from perfect.
- The calls are handled via the function from inside the step. It is not straightforward to make the step "script-function" by itself. This is doable but would need more effort and shaking of how we work with steps in the core.

**Comment by avishniakov** [comment]
To be merged after 0.58.3/0.59.0 is released

**Pull request merged**